### PR TITLE
Add links to relevant docs to README

### PR DESCRIPTION
**Pull request by ben-turner**
The config file documentation can be especially difficult to find. Add a link to that, as well as a couple of helpful sections of the Kubernetes.io website.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,6 +14,12 @@ Kubeadm's scope is limited to the local node filesystem and the Kubernetes API,
 1. **kubeadm upgrade** to upgrade a Kubernetes cluster to a newer version.
 1. **kubeadm reset** to revert any changes made to this host by kubeadm init or kubeadm join.
 
+## Documentation
+
+- [Installation Instructions](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/)
+- [Command Reference](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm/)
+- [Config File Reference](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) (v1beta2)
```
Comment (neolit123) [inline README.md:21]: i'd prefer if we don't reference our WIP (beta) API documentation in yet another location.
potentially you could replace the link with:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/

**Comment by neolit123** [comment]
> The config file documentation can be especially difficult to find.

the documentation is clearly references for both the kubeadm init and join commands:
https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#config-file
https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join/#config-file

**Comment by ben-turner** [comment]
@neolit123, would it be more helpful for me to create an issue with all the things I've had trouble with documentation-wise? I've spent 2 weeks fighting with kubeadm, so I would really like to make it easier for any new users coming through. I know that once you're familiar with a project, everything seems so obvious, but there are a lot of things that make very little sense to someone who is completely new to kubeadm.

As an example, I wouldn't expect to find the config file documentation from a link called "API reference pages."  I'm not trying to write an API integration, I'm trying to write a config file. (Sure, I know now why that's the text for the link, but as a brand new user? That's going to be the last link I click.)

I'm also curious about why you don't want more links to that documentation. Using a config file for certain options is mentioned throughout the docs, but config files are still in beta? It makes it very unclear about what the preferred way of doing things is. It seems to me that either config files need to be fully supported and documented, or there should be very little mention of them in the docs.

**Comment by neolit123** [comment]
> would it be more helpful for me to create an issue with all the things I've had trouble with documentation-wise? I've spent 2 weeks fighting with kubeadm, so I would really like to make it easier for any new users coming through. I know that once you're familiar with a project, everything seems so obvious, but there are a lot of things that make very little sense to someone who is completely new to kubeadm.

you could enumerate the problem areas in a comment in this PR and we can make picks on what to include in this PR or how to improve the k8s.io documentation and potentially get guidance from the k8s documentation maintainers.

> As an example, I wouldn't expect to find the config file documentation from a link called "API reference pages." I'm not trying to write an API integration, I'm trying to write a config file. (Sure, I know now why that's the text for the link, but as a brand new user? That's going to be the last link I click.)

the sections here:
https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#config-file
https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join/#config-file

state that the kubeadm configuration file is a based on a versioned API. for golang projects, it is common for versioned APIs to be documented at godoc reference page. i do not see a problem with that and frankly we have not seen complains about users not clicking the link in the sentence:
>> For more details on each field in the v1beta2 configuration you can navigate to our API reference pages.

> I'm also curious about why you don't want more links to that documentation. Using a config file for certain options is mentioned throughout the docs, but config files are still in beta? It makes it very unclear about what the preferred way of doing things is. It seems to me that either config files need to be fully supported and documented, or there should be very little mention of them in the docs.

there is an underlying structure to our documentation.
"create cluster guide" -> "kubeadm init/join command reference documentation" -> "kubeadm config/API links".

when the latest API changes we modify the relevant pages to include the latest version (e.g. v1beta2).
the API/configuration is still work in progress and i'd prefer if we don't link to the latest API in yet another location.

also note that kubernetes/kubeadm is really not a landing point for new users. kubernetes/kubeadm is an issue tracker that also happens to contain some developer documentation.

minus my comment here https://github.com/kubernetes/kubeadm/pull/2239#pullrequestreview-459570205 the PR seems overall fine.

**Comment by ben-turner** [comment]
> > would it be more helpful for me to create an issue with all the things I've had trouble with documentation-wise? I've spent 2 weeks fighting with kubeadm, so I would really like to make it easier for any new users coming through. I know that once you're familiar with a project, everything seems so obvious, but there are a lot of things that make very little sense to someone who is completely new to kubeadm.
> 
> you could enumerate the problem areas in a comment in this PR and we can make picks on what to include in this PR or how to improve the k8s.io documentation and potentially get guidance from the k8s documentation maintainers.

Excellent. I'll put something here once I've shortened the original essay I wrote since it was way too long.

> > As an example, I wouldn't expect to find the config file documentation from a link called "API reference pages." I'm not trying to write an API integration, I'm trying to write a config file. (Sure, I know now why that's the text for the link, but as a brand new user? That's going to be the last link I click.)
> 
> the sections here:
> https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#config-file
> https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join/#config-file
> 
> state that the kubeadm configuration file is a based on a versioned API. for golang projects, it is common for versioned APIs to be documented at godoc reference page. i do not see a problem with that and frankly we have not seen complains about users not clicking the link in the sentence:
>
> > > For more details on each field in the v1beta2 configuration you can navigate to our API reference pages.

Imagine you are a user, totally unfamiliar with the Go ecosystem and you're trying to write a configuration file for kubeadm. The first thing I would do is run a Google search for "kubeadm config documentation". That gets me to [the kubeadm config](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-config/) page. As I'm scanning through this document, I'm looking for either a block of YAML or a link that says something like "config." Finding neither, I may read more carefully and discover that this is the documentation for a command, not the config itself. Now I poke through the other pages, looking for YAML, a table of keys and descriptions, or a link that says something about configs. At this point maybe I'm on the [init documentation](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/). I'm not looking for information on the command, so I expect most of the information is irrelevant to me, therefore I'm scanning quite quickly for the word "config".

Say I do click [the right link](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) on that page. Now I'm presented with a page that says

> package v1beta2
import "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2"

Yep. Definitely API documentation for developers. Let's go back to the documentation on kubernetes.io.

I'm obviously playing dumb here but it's really not that far off from the experience of someone coming into K8s for the first time. Say I'm a DevOps engineer, never written a line of Go in my life. Management has told me they heard about this great thing called Kubernetes and they want me to set it up. There is a very good chance that I'm going to try to use kubeadm before I'm even familiar with kubernetes itself (beyond maybe reading [the overview](https://kubernetes.io/docs/concepts/overview/what-is-kubernetes/)). There is an even better chance that I don't know all the oddities about the Go and Kubernetes ecosystems with regard to documentation and best practices. Kubeadm is very much one of the first things I see.

> frankly we have not seen complains about users not clicking the link in the sentence

Here you go, your first complaint. [But I really don't think I'm the only person having trouble finding the docs](https://github.com/kubernetes/kubeadm/issues/1251)

> > I'm also curious about why you don't want more links to that documentation. Using a config file for certain options is mentioned throughout the docs, but config files are still in beta? It makes it very unclear about what the preferred way of doing things is. It seems to me that either config files need to be fully supported and documented, or there should be very little mention of them in the docs.
> 
> there is an underlying structure to our documentation.
> "create cluster guide" -> "kubeadm init/join command reference documentation" -> "kubeadm config/API links".

The way these sections are spread out doesn't make the flow totally clear. Let me take a look and see if I can suggest any improvements, perhaps with different "What's Next" links.

> when the latest API changes we modify the relevant pages to include the latest version (e.g. v1beta2).
> the API/configuration is still work in progress and i'd prefer if we don't link to the latest API in yet another location.

Understood.

> also note that kubernetes/kubeadm is really not a landing point for new users. kubernetes/kubeadm is an issue tracker that also happens to contain some developer documentation.

I understand this. The repo that a tool came from seems like a likely place to look for information though, no? That's how I ended up here.

I'm sorry if I'm just putting problems in your lap without solutions or if it seems like I'm just thick as a brick. I just really can't believe that I'm the only one who's been frustrated by this so I'm hoping I can turn that frustration into something that will help others. I'll ask someone else on my team to go through the same setup process and see if they face the same roadblocks I did. That way there's a bit more confirmation that there is actually a problem here and might help identify some better solutions.

**Comment by neolit123** [comment]
you are pointing valid concerns around the organization of the kubeadm docs, search queries and cross linking.
it's important to note that the kubeadm maintainers originally preferred the kubeadm docs to be in a single location at k8s.io, but we ended up having them in separate locations such as "tasks", "setup" and "reference".

i suggest you open a separate issue in kubernetes/website if you wish a discussion with the k8s.io maintainers and try to address the wider problems.

in terms of, this PR, i've already stated what i think is best.
for https://github.com/kubernetes/website/issues/22885, i think https://github.com/kubernetes/website/issues/22885#issuecomment-667513154 should partially resolve the "search engine" problem.

> Yep. Definitely API documentation for developers. Let's go back to the documentation on kubernetes.io.

kubernetes ComponentConfig (i.e. the kubeadm API/config) is documented at godocs, including example usage.
there are exceptions like [this page](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags/), but we have some other plans for it, including documenting support for patches.

**Comment by ben-turner** [comment]
Gotcha. I've updated this PR to follow the points you've mentioned and will move the rest of the discussion over to the website repo once I have some more concrete ideas of how we can improve users' flow through the docs.

File: README.md [hunk 2]
```diff
@@ -14,6 +14,13 @@ Kubeadm's scope is limited to the local node filesystem and the Kubernetes API,
 1. **kubeadm upgrade** to upgrade a Kubernetes cluster to a newer version.
 1. **kubeadm reset** to revert any changes made to this host by kubeadm init or kubeadm join.
 
+## Documentation
+
+- [Installation Instructions](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/)
+- [Creating a single control-plane cluster](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/)
+- [Command Reference](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm/)
+- [Configuration API Reference](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm) - The documentation for each config file version can be found within the [package sub-directories](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm#pkg-subdirectories).
```
Comment (neolit123) [inline README.md:22]: suggesting a minor change here:

> [Configuration API Reference](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm#pkg-subdirectories) (pick an API version from the list of packages)

File: README.md [hunk 3]
```diff
@@ -14,6 +14,13 @@ Kubeadm's scope is limited to the local node filesystem and the Kubernetes API,
 1. **kubeadm upgrade** to upgrade a Kubernetes cluster to a newer version.
 1. **kubeadm reset** to revert any changes made to this host by kubeadm init or kubeadm join.
 
+## Documentation
+
+- [Installation Instructions](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/)
```
Comment (neolit123) [inline README.md:19]: let's use:
> Installing kubeadm

File: README.md [hunk 4]
```diff
@@ -14,6 +14,13 @@ Kubeadm's scope is limited to the local node filesystem and the Kubernetes API,
...
+- [Creating a single control-plane cluster](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/)
+- [Command Reference](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm/)
```
Comment (neolit123) [inline README.md:21]: > Command line reference

**Comment by neolit123** [comment]
Here i suggested a slightly different link /text https://github.com/kubernetes/kubeadm/pull/2239#discussion_r465721611

We can just link to the anchor with the list of packages.

Also please the commits squashed to 1.

**Comment by neolit123** [comment]
squashed the commits in this followup PR:
https://github.com/kubernetes/kubeadm/pull/2276

/close

**Pull request closed without merging**